pkg/stanza/fileconsumer/internal/tracker/tracker.go

```diff
@@ -125,6 +143,34 @@ func (t *fileTracker) TotalReaders() int {
 	return total
 }
 
+func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
+	// We make use of a ring buffer, where each set of files is stored under a specific index.
+	// Instead of discarding knownFiles[2], write it to the next index and eventually roll over.
+	// Separate storage keys knownFilesArchive0, knownFilesArchive1, ..., knownFilesArchiveN, roll over back to knownFilesArchive0
+
+	// Archiving:  ┌─────────────────────on-disk archive─────────────────────────┐
+	//             |    ┌───┐     ┌───┐                     ┌──────────────────┐ |
+	// index       | ▶  │ 0 │  ▶  │ 1 │  ▶      ...       ▶ │ polls_to_archive │ |
+	//             | ▲  └───┘     └───┘                     └──────────────────┘ |
+	//             | ▲    ▲                                                ▼     |
+	//             | ▲    │ Roll over overriting older offsets, if any     ◀     |
+	//             └──────│──────────────────────────────────────────────────────┘
+	//                    │
+	//                    │
+	//                    │
+	//                   start
+	//                   index
+
+	if t.pollsToArchive <= 0 || t.persister == nil {
+		return
+	}
+	key := fmt.Sprintf("knownFiles%d", t.archiveIndex)
+	if err := checkpoint.SaveKey(context.Background(), t.persister, metadata.Get(), key); err != nil {
```

Comment:
Can we just checkpoint in exactly the same place as we were before? I'm confused why we need to add a separate milestone for this.

Reply:
@djaglowski I’ve made some adjustments and relocated the archive checkpointing to just below where we perform our standard checkpointing.

Reply:
I did, but reconsidering this comment - when using an archive, we have to load each set from storage separately, so I see we also have to write it that way.